### PR TITLE
Async configure

### PR DESCRIPTION
**Pull request by dongryphon**

### CODE REVIEW

File: src/_Arguments.js [hunk 1]
```diff
@@ -0,0 +1,205 @@
+"use strict";
```
Comment (israelroldan) [inline src/_Arguments.js]: Rouge file?

File: src/Value.js [hunk 2]
```diff
@@ -188,7 +188,70 @@ class Value extends Item {
...
+                else if ('value' in me) {
+                    converted = me.value + 1;
+                }
+                else {
+                    // No value to adjust...
+                    return 1;
+                }
+
+                ret = 3;
+            }
```
Comment (israelroldan) [inline src/Value.js]: This block above is part of the -vvvv === -v=4 stuff right?

File: src/Cmdlet.js [hunk 3]
```diff
@@ -178,11 +179,19 @@ class Cmdlet {
...
         return this.constructor.switches;
     }
 
+    applyDefaults (params) {
+        this.switches.applyDefaults(params);
+    }
+
+    atRoot () {
+        return !this.parent;
+    }
```
Comment (israelroldan) [inline src/Cmdlet.js]: Funny, I added an isRoot on my local working branch! (good to know we now have one)

File: src/Cmdlet.js [hunk 4]
```diff
@@ -217,6 +266,23 @@ class Cmdlet {
         }
     }
 
+    dispatch (args) {
+        var me = this;
+        var params = me.params;
+
+        args.ownerPush(me);
+
+        return Util.finally(me.configure(args).then(() => {
```
Comment (israelroldan) [inline src/Cmdlet.js]: cool!

File: src/Container.js [hunk 5]
```diff
@@ -26,68 +27,79 @@ class Container extends Cmdlet {
...
+        return args.pull().then(arg => {
+            let counter = me._counter;  // 0 the first time in...
+
+            if (arg) {
+                // After any command has run, the only valid tokens that
+                // should follow are "and" or "then". Unless this is the
+                // root container, in which case we allow commands to be
+                // run in an list:
+                //
+                //      cmd foo bar baz
```
Comment (israelroldan) [inline src/Container.js]: good 'ol "ant clean build" syntax but instead of target names, commands. neat!

File: src/Help.js [hunk 6]
```diff
@@ -97,23 +99,17 @@ class Help extends Command {
...
+                    out.push(` * ${command.name}${command.type.help ? ':\t' + command.type.help + '\t' : ''}${command.aliases.length > 0 ? ' (or ' + command.aliases.join(', ') + ')': ''}`);
                 });
                 syntaxTokenParts.push('subcommand');
             }
 
             // This container has additional containers and/or commands, push the placeholder into the syntax string
-            syntax.push(`${defaultCmd ? '(' : '[' }${syntaxTokenParts.join('|')}${defaultCmd ? ')\tExecute subcommand' : ']' }`);
+            if (syntaxTokenParts.length) {
+                syntax.push(`[${syntaxTokenParts.join('|')}]`);
+            }
```
Comment (israelroldan) [inline src/Help.js]: The idea behind the removed code was to show something like this:

    Syntax:
        git remote                                  Show a list of remotes
        git remote [options] [command]              Execute a command

If we just show the latter (with no "execute a command", of course), how can users know that running that command does something without subcommand? (I have a better version of the help in the works which presents this in a better way)

File: src/Items.js [hunk 7]
```diff
@@ -139,7 +165,8 @@ class Items {
     canAdd (item) {
         let name = item.name;
 
-        if (this.map[name] || this.map[item.loname]) {
+        // Allow derived classes to overwrite inherited items
+        if (this.map.hasOwnProperty(name) || this.map.hasOwnProperty(item.loname)) {
             throw new Error(`Duplicate ${this.kind} "${name}"`);
```
Comment (israelroldan) [inline src/Items.js]: This was on my question list on our last call but completely forgot about it!

File: src/Help.js (hunk 6 above)
Comment (dongryphon) [inline src/Help.js]: Makes sense... the issue was the name "default" originally was put into the text... now that is blank so :p but agreed we should display info of some kind on the default command

File: src/Value.js (hunk 2 above)
Comment (dongryphon) [inline src/Value.js]: Yes... the block below for numbers is to support the eventual equivalence of `-vvvv` and `--v=4` (assuming the param value started at `0` of course)

File: src/Help.js (hunk 6 above)
Comment (israelroldan) [inline src/Help.js]: cool! It's coming back soon, then! :)

**Pull request merged**